### PR TITLE
doc: validator structure change, state witness reference impl

### PR DESCRIPTION
**Pull request by Longarithm**
Documenting changes to validators and describing basics of reference implementation.

### CODE REVIEW

File: neps/nep-0509.md [hunk 1]
```diff
@@ -57,32 +57,45 @@ As a result, the team sought alternative approaches and concluded that stateless
 * More shards - this is covered in the resharding project.
 * ZK integration.
 * Underlying data structure change (e.g. verkle tree).
-* Change to validator rewards.
 * TBD
 
 ## High level flow
 
-We propose a change to the following parts of the chunk and block production flow:
+The current high-level chunk production flow, if we drop details and edge cases, is as follows:
```
Comment (pugachAG) [inline neps/nep-0509.md:64]: just FYI chat gpt suggested the following when I asked to make this sentence more idiomatic: 
```
The current high-level chunk production flow, excluding details and edge cases, is as follows
```

File: neps/nep-0509.md [hunk 2]
```diff
@@ -57,32 +57,45 @@ As a result, the team sought alternative approaches and concluded that stateless
...
+* Block producer at height H BP(H) produces block B(H) with chunks accessible to it and distributes it.
```
Comment (pugachAG) [inline neps/nep-0509.md:65]: nit: I suggest highlighting expressions as code to make it easier to read: 
Block producer at height `H` `BP(H)` produces block `B(H)`...

File: neps/nep-0509.md [hunk 3]
```diff
@@ -47,135 +48,368 @@ As a result, the team sought alternative approaches and concluded that stateless
...
-We also propose a change to the validator roles and responsibilities. This is the list of roles after the proposal, with same and new behavior clearly labelled:
+So the high-level specification can be described as the list of changes in the validator roles and responsibilities:
 
 * Block producers:
   * (Same as today) Produce blocks, (new) including waiting for chunk endorsements
   * (Same as today) Maintain chunk parts (i.e. participates in data availability based on Reed-Solomon erasure encoding)
-  * (Same as today) Do not require tracking any shard
-  * (Same as today) Should have a higher barrier of entry for security reasons (e.g. to make block double signing harder)
+  * (New) No longer require tracking any shard
+  * (Same as today) Should have a higher barrier of entry (to keep `BlockHeader` size low and for security reasons, to make block double signing harder)
```
Comment (pugachAG) [inline neps/nep-0509.md:85]: how does " higher barrier of entry" helps with "to keep `BlockHeader` size low"? 🤔

File: neps/nep-0509.md [hunk 4]
```diff
@@ -47,135 +48,368 @@ As a result, the team sought alternative approaches and concluded that stateless
...
+Reward for each validator is defined as `total_epoch_reward * validator_relative_stake * work_quality_ratio`, where:
+* `total_epoch_reward` is selected so that total inflation of the token is 5% per annum;
+* `validator_relative_stake = validator_stake / total_epoch_stake`;
+* `work_quality_ratio` is the measure of the work quality from 0 to 1.
+
+So, the actual reward never exceeds total reward, and when everyone does perfect work, they are equal.
+For the context of the NEP, it is enough to assume that `work_quality_ratio = avg_{role}({role}_quality_ratio)`.
+So, if node is both a block and chunk producer, we compute quality for each role separately and then take average of them.
+  
+When epoch is finalized, all headers of blocks in it uniquely determine who was expected to produce each block and chunk.
```
Comment (pugachAG) [inline neps/nep-0509.md:167]: nit: `headers of blocks` -> `block headers`

File: neps/nep-0509.md [hunk 5]
```diff
@@ -47,135 +48,368 @@ As a result, the team sought alternative approaches and concluded that stateless
...
+    /// After these are applied as well, we should arrive at the pre-state-root
+    /// of the chunk that this witness is for.
+    pub implicit_transitions: Vec<ChunkStateTransition>,
+```
+
+Then, while our shard was missing chunks, other shards could still produce chunks, which could generate receipts targeting our shards. So, we need to extend `source_receipt_proofs`. 
+Field structure doesn't change, but we need to carefully pick range of set of source chunks, so different subsets will cover all source receipts without intersection.
+
+Let's say B2 is the block that contains the last new chunk of shard S before chunk which state transition we execute, and B1 is the
+block that contains the last new chunk of shard S before B2.
```
Comment (pugachAG) [inline neps/nep-0509.md:309]: nit: you don't need line break here

**Review by pugachAG — APPROVED** [review]

**Comment by wacban** [comment]
Sorry I won't have time to review today, feel free to merge to reduce the chance of merge conflicts with others. If I have any nits I'll just send a follow up PR.

**Pull request merged**